pkg/common/vsphere/virtual_center.go

```diff
@@ -0,0 +1,289 @@
+		return nil, err
+	}
+	return vslmClient, nil
+}
+
+// NewCnsClient creates a new CNS client
+func newCnsClient(ctx context.Context, c *vim25.Client, logger logrus.FieldLogger) (*cns.Client, error) {
+	cnsClient, err := cns.NewClient(ctx, c)
+	if err != nil {
+		logger.Errorf("failed to create a new client for CNS. err: %v", err)
+		return nil, err
+	}
+	return cnsClient, nil
+}
+
+// newClient creates a new govmomi Client instance.
+func (vc *VirtualCenter) newClient(ctx context.Context) (*govmomi.Client, error) {
+	log := vc.logger
+	if vc.Config.Scheme == "" {
+		vc.Config.Scheme = DefaultScheme
+	}
+
+	url, err := soap.ParseURL(net.JoinHostPort(vc.Config.Host, strconv.Itoa(vc.Config.Port)))
+	if err != nil {
+		log.Errorf("failed to parse URL %s with err: %v", url, err)
+		return nil, err
+	}
+	//url.Path = "/sdk"
+	soapClient := soap.NewClient(url, vc.Config.Insecure)
+	soapClient.Timeout = time.Duration(vc.Config.VCClientTimeout) * time.Minute
+	log.Debugf("Setting vCenter soap client timeout to %v", soapClient.Timeout)
+	vimClient, err := vim25.NewClient(ctx, soapClient)
+	if err != nil {
+		log.Errorf("failed to create new client with err: %v", err)
+		return nil, err
+	}
+	vimClient.RoundTripper = session.KeepAlive(vimClient.RoundTripper, 10*time.Minute)
+	err = vimClient.UseServiceVersion("vsan")
+	if err != nil && vc.Config.Host != "127.0.0.1" {
+		// skipping error for simulator connection for unit tests
+		log.Errorf("Failed to set vimClient service version to vsan. err: %v", err)
+		return nil, err
+	}
+	vimClient.UserAgent = "cns-dp-useragent"
+
+	client := &govmomi.Client{
+		Client:         vimClient,
+		SessionManager: session.NewManager(vimClient),
+	}
+
+	err = vc.login(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := client.SessionManager.UserSession(ctx)
+	if err == nil {
+		log.Infof("New session ID for '%s' = %s", s.UserName, s.Key)
+	}
+
+	if vc.Config.RoundTripperCount == 0 {
+		vc.Config.RoundTripperCount = DefaultRoundTripperCount
+	}
+	client.RoundTripper = vim25.Retry(client.RoundTripper, vim25.TemporaryNetworkError(vc.Config.RoundTripperCount))
+	return client, nil
+}
+
+// SessionManager.Login with user and password.
+func (vc *VirtualCenter) login(ctx context.Context, client *govmomi.Client) error {
+	log := vc.logger
+	log.Infof("Attempting login for user: %s", vc.Config.Username)
+	err := client.SessionManager.Login(ctx, neturl.UserPassword(vc.Config.Username, vc.Config.Password))
+	if err != nil {
+		log.Errorf("Failed for login for username : %s error: %v", vc.Config.Username, err)
+	}
+	return err
+}
+
+// Connect establishes a new connection with vSphere with updated credentials
+// If credentials are invalid then it fails the connection.
+func (vc *VirtualCenter) Connect(ctx context.Context) (*vslm.GlobalObjectManager, *cns.Client, error) {
+	vc.connectMutex.Lock()
+	defer vc.connectMutex.Unlock()
+	log := vc.logger
+	// Set up the vc connection
+	err := vc.connect(ctx, false)
```

Comment:
I don't see the internal routine being called anywhere else.  Let's squash this into Connect.